testing/hyperparams/test_space.py

```diff
@@ -89,3 +90,35 @@ def test_hyperparams_space_rvs_outputs_samples():
     for k, v in samples.iter_flat():
         assert k in space
         assert not isinstance(v, HyperparameterDistribution)
+
+
+def test_hyperparameter_samples_compress():
+    _flat_hps = {
+        "b__a__learning_rate": 7,
+        "b__learning_rate": 9,
+        "Sklearn__test1__test__abc": False,
+        "Sklearn__test2__test__abc": "Parallel"
+    }
+    expected_compressed_hps_with_parents = [
+        {'step_name': 'a', 'hyperparams': {'learning_rate': 7}, 'ancestor_steps': ['b']},
+        {'step_name': 'b', 'hyperparams': {'learning_rate': 9}, 'ancestor_steps': []},
+        {'step_name': 'test', 'hyperparams': {'abc': False}, 'ancestor_steps': ["Sklearn", "test1"]},
+        {'step_name': 'test', 'hyperparams': {'abc': "Parallel"}, 'ancestor_steps': ["Sklearn", "test2"]}]
+    expected_compressed_hps_with_out_parents = [
+        {'step_name': 'a', 'hyperparams': {'learning_rate': 7}, 'ancestor_steps': None},
+        {'step_name': 'b', 'hyperparams': {'learning_rate': 9}, 'ancestor_steps': None},
+        {'step_name': 'test', 'hyperparams': {'abc': False}, 'ancestor_steps': None},
+        {'step_name': 'test', 'hyperparams': {'abc': "Parallel"}, 'ancestor_steps': None}]
+    expected_compressed_wild_cards = [('*a__learning_rate', 7), ('b__learning_rate', 9), ('*test1*abc', False),
+                                      ("*test2*abc", 'Parallel')]
+    _separator = "__"
+    hps = HyperparameterSamples(_flat_hps, separator=_separator)
+    hps_compressed_with_parents = hps.compress()
+    assert isinstance(hps_compressed_with_parents, CompressedHyperparameterSamples)
+    assert str(hps_compressed_with_parents) == str(expected_compressed_hps_with_parents)
+
+    hps_compressed_without_parents = hps.compress(remove_parents=True)
+    assert isinstance(hps_compressed_without_parents, CompressedHyperparameterSamples)
+    assert str(hps_compressed_without_parents) == str(expected_compressed_hps_with_out_parents)
+
+    assert list(hps_compressed_with_parents.wildcards().items()) == expected_compressed_wild_cards
```

Comment:
Your unit test currently follows this sequence: 
```python
# arrange
# arrange
# act
# assert
# act
# assert
# assert
```

However, please see [here](https://medium.com/@pjbgf/title-testing-code-ocd-and-the-aaa-pattern-df453975ab80#:~:text=The%20AAA%20(Arrange%2DAct%2D,%3A%20arrange%2C%20act%20and%20assert.) that unit tests should be only like this: 

```python
# arrange
# act
# assert
```

It looks like your unit test could be separated in at least three different tests: 
1. test_hyperparameter_samples_compress_with_parents
2. test_hyperparameter_samples_compress_without_parents
3. test_hyperparameter_samples_compress_wildcards

Also, it is often good to have more tests. The wildcards algo is complicated and will probably require another test (so 4 tests). Could you think of more complicated test cases to try the algorithm on ?

Reply:
Yes sure, i will try different pattern's to see if there is any testcase at which the algo fails.

Reply:
Oh, this `aaa` unit test pattern feels pretty much super efficient. Will definitely incorporate them in the PR.